src/ripple/core/impl/DeadlineTimer.cpp

```diff
@@ -59,11 +61,13 @@ class DeadlineTimer::Manager
     // However, an extra notification may still happen due to concurrency.
     //
     void activate (DeadlineTimer& timer,
-        double secondsRecurring, beast::RelativeTime const& when)
+        duration recurring,
+        time_point when)
     {
-        assert (secondsRecurring >= 0);
+        using namespace std::chrono_literals;
+        assert (recurring >= 0ms);
 
-        std::lock_guard <std::recursive_mutex> lock (m_mutex);
+        std::lock_guard <std::recursive_mutex> lock {mutex_};
```

Comment:
Good question.  It already was when I got here.  My strong suspicion is that, yes it needs to be recursive.

Here's the scenario:
1. First, note that the callback happens while the lock is held.
2. If, during the callback, the callee were to call the DeadlineTimer to make an adjustment, then it would be a locking call on the same thread.

Assuming we want to support this scenario, then I believe the mutex needs to be recursive.